### PR TITLE
Set up Github actions for publishing a release

### PR DESCRIPTION
**Pull request by Egorand**

### CODE REVIEW

**Review by oldergod — COMMENTED** [review]
What do we do about the current publish tasks?

https://github.com/square/okio/blob/master/.github/workflows/build.yml#L74-L160

**Comment by Egorand** [comment]
We still need those for publishing snapshots. Perhaps there's a way to organize the jobs better without duplication?

File: .github/workflows/publish-release.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+jobs:
+  create_staging_repository:
+    runs-on: ubuntu-latest
+    name: Create staging repository
+    outputs:
+      repository_id: ${{ steps.create.outputs.repository_id }}
+    steps:
+      - id: create
+        uses: nexus-actions/create-nexus-staging-repo@v1.1
```
Comment (martinbonnin) [inline .github/workflows/publish-release.yml:18]: You might want to replace the version with its sha1 to make sure the tag isn't moved under the hood:

```suggestion
        uses: nexus-actions/create-nexus-staging-repo@10a43a22d0278ed006f1293214b99184c60fbdff
```

I am a maintainer of this action and I can assure you there is nothing fishy going on down there but you shouldn't trust me :)

File: .github/workflows/publish-release.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          ORG_GRADLE_PROJECT_signingInMemoryKey: ${{ secrets.ARTIFACT_SIGNING_PRIVATE_KEY }}
+
+  finalize:
+    runs-on: ubuntu-latest
+    needs: [create_staging_repository, publish, publish_windows]
+    if: ${{ always() && needs.create_staging_repository.result == 'success' }}
+    steps:
+      - name: Discard
+        if: ${{ needs.publish.result != 'success' || needs.publish_windows.result != 'success' }}
+        uses: nexus-actions/drop-nexus-staging-repo@v1
```
Comment (martinbonnin) [inline .github/workflows/publish-release.yml:80]: Same here
```suggestion
        uses: nexus-actions/drop-nexus-staging-repo@fe83783967a063540320ace3c8942608246705a1
```

File: .github/workflows/publish-release.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        with:
+          username: ${{ secrets.SONATYPE_NEXUS_USERNAME }}
+          password: ${{ secrets.SONATYPE_NEXUS_PASSWORD }}
+          staging_repository_id: ${{ needs.create_staging_repository.outputs.repository_id }}
+      - name: Release
+        if: ${{ needs.publish.result == 'success' && needs.publish_windows.result == 'success' }}
+        uses: nexus-actions/release-nexus-staging-repo@v1.2
```
Comment (martinbonnin) [inline .github/workflows/publish-release.yml:87]: Same here

```suggestion
        uses: nexus-actions/release-nexus-staging-repo@36161f25ef98cc3821eabb11ab742d2e9d479e52
```

File: .github/workflows/publish-release.yml (hunk 1 above)
Comment (Egorand) [reply .github/workflows/publish-release.yml:18]: Nice, thanks! Not super familiar with Github actions versioning.

File: .github/workflows/publish-release.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+jobs:
+  create_staging_repository:
+    runs-on: ubuntu-latest
+    name: Create staging repository
+    outputs:
+      repository_id: ${{ steps.create.outputs.repository_id }}
+    steps:
+      - id: create
+        uses: nexus-actions/create-nexus-staging-repo@10a43a22d0278ed006f1293214b99184c60fbdff // v1.1
```
Comment (swankjesse) [inline .github/workflows/publish-release.yml:18]: sweeet

File: .github/workflows/publish-release.yml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+          ORG_GRADLE_PROJECT_mavenCentralUsername: ${{ secrets.SONATYPE_NEXUS_USERNAME }}
+          ORG_GRADLE_PROJECT_mavenCentralPassword: ${{ secrets.SONATYPE_NEXUS_PASSWORD }}
+          ORG_GRADLE_PROJECT_signingInMemoryKey: ${{ secrets.ARTIFACT_SIGNING_PRIVATE_KEY }}
+
+  finalize:
+    runs-on: ubuntu-latest
+    needs: [create_staging_repository, publish, publish_windows]
+    if: ${{ always() && needs.create_staging_repository.result == 'success' }}
+    steps:
+      - name: Discard
```
Comment (swankjesse) [inline .github/workflows/publish-release.yml:78]: Nice

**Review by swankjesse — APPROVED** [review]

**Comment by swankjesse** [comment]
Next steps on this? I’m inclined to consider this bitrotted?

**Comment by Egorand** [comment]
@swankjesse I'd love to get it to work and will try to find time. The remaining bits are locating the correct `staging_profile_id` and bumping all Actions this setup uses to latest versions, and then let's merge it and test? We'd need to publish a few fake releases to ensure this works, hope it's fine.

**Comment by swankjesse** [comment]
I think Vanniktech will do it all now.
https://github.com/square/okio/pull/1201

**Pull request closed without merging**